### ISSUE TITLE
Rule proposal: prevent usage multiple render methods

### ISSUE THREAD

**Issue by LeoVS09** [yannickcr/eslint-plugin-react#2552]
JSX used for giving a declarative style of composing multiple components, but with the usage of multiple render methods it becomes completely useless.
like:
```jsx
class Component extends React.Component {

  // ...

  render(){
    return (
      <div>
        {this.renderHeader()}
        {this.renderBody()}
        {this.renderFooter()}
      </div>
    )
  }
}
```

When better write
```jsx
class Component extends React.Component {
  render(){
    return (
      <div>
       <Header />
       <Body />
       <Footer />
      </div>
    )
  }
}
```
**which also require from developer write more reusable fragments**

The proposal is to mark usage JSX in all other methods except simple `render` as a warning

Example with the **warning**:
```jsx
class Component extends React.Component {

 renderSomething(){
  return (
   <Something />
  )
 }

 render(){
  return (
   <div>
    {this.renderSomething()}
   </div>
  )
 }
}
```

Example with **valid** code:

```jsx
class Component extends React.Component {
 render(){
  return (
   <div>
    <Something />
   </div>
  )
 }
}
```

**Comment by fabiendem** [comment]
Can be useful: https://github.com/sebastian-software/eslint-plugin-shopify-lean/blob/master/docs/rules/react-no-multiple-render-methods.md

**Comment by ljharb** [comment]
I'm not sure this rule would have a lot of value.

Many components are moving to SFCs with hooks; and it's a very subjective discussion about whether things should be in render methods or separate components.

**Comment by fabiendem** [comment]
Still wrapping my head around this.

> Many components are moving to SFCs with hooks

Yeah, I didn't look at the example in the first message. 
Either way, you can do this in functional components as well:

```tsx
export default function MyComponent(props: Props) {
  
  function renderHeader(...) {
    return (
      <View ..../>
    );
  }

  function renderContent(bar) {
    return (
        <View> .....</View>
    );
  }

  return (
    <View>
    {renderHeader(foo, bar)}
    {renderContent(bar)}
    </View>
  );
```

This article explains better why you would want such a rule, or at least avoid the pattern: https://kentcdodds.com/blog/dont-call-a-react-function-component
But I agree, it can get subjective.
